Offboard mode: Fix errors and make it less MAVLink-centric (#2334)
PX4/PX4-user_guide: en/flight_modes/offboard.md

In order to hold position in this case the vehicle must receive a stream of setpoints for the current position.

When using ROS2 the proof that the external source is alive is provided by a stream of [OffboardControlMode](../msg_docs/OffboardControlMode.md) messages, while the actual setpoint is provided by publishing to one of the setpoint uORB topics, such as [TrajectorySetpoint](../en/msg_docs/TrajectorySetpoint.md).
In order to hold position in this case the vehicle must receive a stream of `OffboardControlMode` but would only need the `TrajectorySetpoint` once.

Note that offboard mode only supports only a very limited set of MAVLink commands and messages.
Operations, like taking off, landing, return to launch, may be best handled using the appropriate modes. 
Operations like uploading, downloading missions can be performed in any mode.

## ROS 2 Messages

The following ROS 2 messages and their particular fields and field values are allowed for the specified frames.
In addition to providing heartbeat functionality, `OffboardControlMode` has two other main purposes:

> 1. Controls at which level of the [PX4 control architecture](../en/flight_stack/controller_diagrams) the offboard setpoints must be injected and disables the bypassed controllers.

Author: ```suggestion
1. Controls the level of the [PX4 control architecture](../en/flight_stack/controller_diagrams.md) at which offboard setpoints must be injected, and disables the bypassed controllers.

```

When using ROS2 the proof that the external source is alive is provided by a stream of [OffboardControlMode](../msg_docs/OffboardControlMode.md) messages, while the actual setpoint is provided by publishing to one of the setpoint uORB topics, such as [TrajectorySetpoint](../en/msg_docs/TrajectorySetpoint.md).
In order to hold position in this case the vehicle must receive a stream of `OffboardControlMode` but would only need the `TrajectorySetpoint` once.

Note that offboard mode only supports only a very limited set of MAVLink commands and messages.
Operations, like taking off, landing, return to launch, may be best handled using the appropriate modes. 
Operations like uploading, downloading missions can be performed in any mode.

## ROS 2 Messages

The following ROS 2 messages and their particular fields and field values are allowed for the specified frames.
In addition to providing heartbeat functionality, `OffboardControlMode` has two other main purposes:

1. Controls the level of the [PX4 control architecture](../en/flight_stack/controller_diagrams.md) at which offboard setpoints must be injected, and disables the bypassed controllers.

> 1. Determines which valid estimates (position or velocity) are required.

Author: ```suggestion
1. Determines which valid estimates (position or velocity) are required, and also which setpoint messages should be used.

```

The `OffboardControlMode` message is defined as shown.

```
# Off-board control mode

uint64 timestamp		# time since system start (microseconds)

bool position
bool velocity
bool acceleration
bool attitude
bool body_rate
bool actuator
```

> The fields are ordered in terms of priority such that `position` takes precedence over all the other, `velocity` takes precedence over `acceleration` and the other below it, etc.
> Depending on the first non-zero value from top to bottom, the valid estimate required and the right setpoint message(s) are defined.

Author: ```suggestion
The fields are ordered in terms of priority such that `position` takes precedence over `velocity` and later fields, `velocity` takes precedence over `acceleration`, and so on.
The first field that has a non-zero value (from top to bottom) defines what valid estimate is required in order to use offboard mode, and the setpoint message(s) that can be used.
For example, if the `acceleration` field is the first non-zero value, then PX4 requires a valid `velocity estimate`, and the setpoint must be specified using the `TrajectorySetpoint` message.

```

> | desired control quantity | position field | velocity field | acceleration field | attitude field | body_rate field | acturator field | required estimate | required message                                                                                                                |

Author: acturator - a typo?

Reviewer: Yes, sorry!

bool acceleration
bool attitude
bool body_rate
bool actuator
```

The fields are ordered in terms of priority such that `position` takes precedence over `velocity` and later fields, `velocity` takes precedence over `acceleration`, and so on.
The first field that has a non-zero value (from top to bottom) defines what valid estimate is required in order to use offboard mode, and the setpoint message(s) that can be used.
For example, if the `acceleration` field is the first non-zero value, then PX4 requires a valid `velocity estimate`, and the setpoint must be specified using the `TrajectorySetpoint` message.

| desired control quantity | position field | velocity field | acceleration field | attitude field | body_rate field | acturator field | required estimate | required message                                                                                                                |
|--------------------------|:--------------:|:--------------:|:------------------:|:--------------:|:---------------:|:---------------:|:-----------------:|---------------------------------------------------------------------------------------------------------------------------------|

> | position (NED)           |     &check;    |        -       |          -         |        -       |        -        |        -        |      position     | TrajectorySetpoint                                                                                                              |
> | velocity (NED)           |     &cross;    |     &check;    |          -         |        -       |        -        |        -        |      velocity     | TrajectorySetpoint                                                                                                              |
> | acceleration (NED)       |     &cross;    |     &cross;    |       &check;      |        -       |        -        |        -        |      velocity     | TrajectorySetpoint                                                                                                              |

Author: ```suggestion
| position (NED)           |     &check;    |        -       |          -         |        -       |        -        |        -        |      position     | `TrajectorySetpoint`                                                                                                              |
| velocity (NED)           |     &cross;    |     &check;    |          -         |        -       |        -        |        -        |      velocity     | `TrajectorySetpoint`                                                                                                              |
| acceleration (NED)       |     &cross;    |     &cross;    |       &check;      |        -       |        -        |        -        |      velocity     | `TrajectorySetpoint`                                                                                                              |

```

| thrust and torque (RFD)  |     &cross;    |     &cross;    |       &cross;      |     &cross;    |     &cross;     |     &check;     |        none       | [VehicleThrustSetpoint](../msg_docs/VehicleThrustSetpoint.md) and [VehicleTorqueSetpoint](../msg_docs/VehicleTorqueSetpoint.md) |

where &check; means that the bit is set, &cross; means that the bit is not set and `-` means that the bit is value is irrelevant.

### Copter

* [px4_msgs::msg::TrajectorySetpoint](https://github.com/PX4/PX4-Autopilot/blob/main/msg/TrajectorySetpoint.msg)
  * The following input combinations are supported:
    * Position setpoint (`position` different from `NaN`). Non-`NaN` values of velocity and acceleration are used as feedforward terms for the inner loop controllers.
    * Velocity setpoint (`velocity` different from `NaN` and `position` set to `NaN`). Non-`NaN` values acceleration are used as feedforward terms for the inner loop controllers.

>     * Acceleration setpoint  (`acceleration` different from `NaN` and `position` and `velocity` set to `NaN`)

Author: ```suggestion
    * Acceleration setpoint (`acceleration` different from `NaN` and `position` and `velocity` set to `NaN`)

```

  * The following input combination is supported:
    * quaterion `q_d` + thrust setpoint `thrust_body`. Non-`NaN` values of `yaw_sp_move_rate` are used as feedforward terms expressed in Earth frame and in \[rad/s\].
  - The quaterion represents the rotation between the drone body FRD (front, right, down) frame and the NED frame. The trust is in the drone body FRD frame and expressed in normalized \[-1, 1\] values.

* [px4_msgs::msg::VehicleRatesSetpoint](https://github.com/PX4/PX4-Autopilot/blob/main/msg/VehicleRatesSetpoint.msg)
  * The following input combination is supported:
    * `roll`, `pitch`, `yaw` and `thrust_body`.
  - All the value are in the drone body FRD frame. The rates are in \[rad/s\] while thrust_body is normalized in \[-1, 1\].

* [px4_msgs::msg::VehicleThrustSetpoint](https://github.com/PX4/PX4-Autopilot/blob/main/msg/VehicleThrustSetpoint.msg) + [px4_msgs::msg::VehicleTorqueSetpoint](https://github.com/PX4/PX4-Autopilot/blob/main/msg/VehicleTorqueSetpoint.msg)
  * The following input combination is supported:
    * `xyz` for thrust and `xyz` for torque.
  - All the value are in the drone body FRD frame and normalized in \[-1, 1\].

> 

Author: ```suggestion
```
